### PR TITLE
make ctest actually verify dumbinputs{f} against ref output, fixes #253

### LINKED ISSUE #253: cmake test should diff check the terminal outputs for testutils/dumbinputs
We need to add to the CMake / ctest a step that diffs the *.out against the *.refout
Grep for `diff` in my old `test/check_finufft.sh`

Currently the ctest merely runs testutils, dumbinputs{f}   but does nothing to check their output, which is their point.

Tweaks to add diffs to these ctest commands would be needed:
https://github.com/flatironinstitute/finufft/blob/a0db2d0b2ce275f450095a1007208d21a52a9be9/test/CMakeLists.txt#L25
https://github.com/flatironinstitute/finufft/blob/a0db2d0b2ce275f450095a1007208d21a52a9be9/test/CMakeLists.txt#L53

This is important for long-term CI since it tests various edge cases.

The question is if `diff` is available on windows - @DiamonDinoia Marco or @lu1and10 maybe you know this?

### PR DESCRIPTION
**Pull request by ahbarnett**
If CI is to use cmake, it actually has to test a few things, like this one.
I am new to cmake/ctest, but this was my attempt, after hours lost in docs and stackoverflow. Scripting in ctest is not intuitive. Please improve it! It works, but dependencies may not have been done the best.

You can ignore the file in devel/ which was a discarded experiment, not currently used.

We should then do same for testutils.

My main struggles are:
* is the pipe > platform-independent?
* How can I make dumbinputs{f} run at ctest time not at build time like it does now?
* can OMP_NUM_THREADS be set in platform-indep way?   The basic tests run slowly if all threads given to them, due to thread collisions. Thus I want a small number here.

THe whole point of this cmake pain is to be platform-independent, yet it's very difficult to massage standard scripting tricks into cmake-platform-indep form, or even to know *what* commands are platform-indep...

### CODE REVIEW

**Comment by ahbarnett** [comment]
I would be interested what @DiamonDinoia suggests here.

**Comment by lu1and10** [comment]
> If CI is to use cmake, it actually has to test a few things, like this one. I am new to cmake/ctest, but this was my attempt, after hours lost in docs and stackoverflow. Scripting in ctest is not intuitive. Please improve it! It works, but dependencies may not have been done the best.
> 
> You can ignore the file in devel/ which was a discarded experiment, not currently used.
> 
> We should then do same for testutils.
> 
> My main struggles are:
> 
>     * is the pipe > platform-independent?
> 
>     * How can I make dumbinputs{f} run at ctest time not at build time like it does now?
> 
>     * can OMP_NUM_THREADS be set in platform-indep way?   The basic tests run slowly if all threads given to them, due to thread collisions. Thus I want a small number here.
> 
> 
> THe whole point of this cmake pain is to be platform-independent, yet it's very difficult to massage standard scripting tricks into cmake-platform-indep form, or even to know _what_ commands are platform-indep...

That's what I will do too. The add_test command only accepts one executable, you have to make a wrapper script `check_dumbinputs.cmake` to actually call the two commands sequentially. Not sure if there is other neater way in cmake...

**Comment by lu1and10** [comment]
> If CI is to use cmake, it actually has to test a few things, like this one. I am new to cmake/ctest, but this was my attempt, after hours lost in docs and stackoverflow. Scripting in ctest is not intuitive. Please improve it! It works, but dependencies may not have been done the best.
> 
> You can ignore the file in devel/ which was a discarded experiment, not currently used.
> 
> We should then do same for testutils.
> 
> My main struggles are:
> 
>     * is the pipe > platform-independent?
> 
>     * How can I make dumbinputs{f} run at ctest time not at build time like it does now?
> 
>     * can OMP_NUM_THREADS be set in platform-indep way?   The basic tests run slowly if all threads given to them, due to thread collisions. Thus I want a small number here.
> 
> 
> THe whole point of this cmake pain is to be platform-independent, yet it's very difficult to massage standard scripting tricks into cmake-platform-indep form, or even to know _what_ commands are platform-indep...

I think the following diff should also work, so that you can remove `add_custom_target` avoiding the redirection `>` 
```
diff --git a/devel/check_dumbinputs.cmake b/devel/check_dumbinputs.cmake
index f503a16..220fd33 100644
--- a/devel/check_dumbinputs.cmake
+++ b/devel/check_dumbinputs.cmake
@@ -8,7 +8,7 @@ set(tempout
 
 # pipe dumbinputs there. Is it platform-indep?
 # https://stackoverflow.com/questions/36304289/how-to-use-redirection-in-cmake-add-test
-execute_process(COMMAND ${CMAKE_CURRENT_BINARY_DIR}/dumbinputs > ${tempout})
+execute_process(COMMAND ${CMAKE_CURRENT_BINARY_DIR}/dumbinputs OUTPUT_FILE ${tempout})
 
 # diff the output against reference
 execute_process(COMMAND ${CMAKE_COMMAND} -E compare_files ${tempout} ${CMAKE_CURRENT_SOURCE_DIR}/results/dumbinputs.refout)
diff --git a/test/CMakeLists.txt b/test/CMakeLists.txt
index 2a88c4f..cb5c455 100644
--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -74,7 +74,7 @@ add_custom_target(
 # See: https://stackoverflow.com/questions/3065220/ctest-with-multiple-commands?rq=4
 add_test(
   NAME check_dumbinputs_${PREC}
-  COMMAND ${CMAKE_COMMAND} -E compare_files ${tempout} ${CMAKE_CURRENT_SOURCE_DIR}/results/dumbinputs${SUFFIX}.refout
+  COMMAND ${CMAKE_COMMAND} -P ${CMAKE_SOURCE_DIR}/devel/check_dumbinputs.cmake
   )
 
 endfunction()
```

**Comment by wendazhou** [comment]
I concur with @DiamonDinoia , I would strongly recommend writing the full test in C++. Scripting is not really platform independent (i.e. shell is a unix concept, different shells have different syntax and built-in utilities etc.). I'm also pretty sure the cmake code above will break with multi-config generators such as [ninja multi-config](https://cmake.org/cmake/help/latest/generator/Ninja%20Multi-Config.html) or visual studio.

In this specific case, I would suggest to modify the test in the following fashion:
add a macro to handle the assertion, e.g.
```{cpp}
#define ASSERT_ERROR(MSG, X) { printf(MSG " (should complain)\n", X); if(!X) { return 1; } }
```
and change the test as
```{cpp}
 int ier = FINUFFT1D1(M,x,c,+1,0,N,F,&opts);
 // printf("1d1 tol=0:\tier=%d (should complain)\n",ier);
ASSERT_ERROR("1d1 tol=0:\tier=%d", ier);
```
This allows you to just run the executable through `add_test`. You will also avoid the need for brittle text comparisons (prone to newline issues across platforms), and better encode the test expectations

**Comment by blackwer** [comment]
Agree about newlines and other perils in scripting. There is a workaround for that, but it just makes the scripting even more cumbersome: copying the files with the cmake copy and translating the line endings. 
That said, if it works, it works. multibuilds are edge case enough that breaking them elicits a big _shrug_ from me.

I pretty much always use C++ tests. Python theoretically removes platform dependence, but it's really only good when you're requiring python in the build environment in the first place. Otherwise you're yanking in a python dependency just to build tests.

**Comment by ahbarnett** [comment]
ctest/cmake was so horrible that it was worth taking Wenda's suggestion and baking all checks into the C++ code. It is probably more clear that way - felt like cleaning out the back of some long-forgotten cupboard :)

**Comment by ahbarnett** [comment]
THis means we can use this ctest in the CI... next step.

**Comment by ahbarnett** [comment]
This issue remains: can OMP_NUM_THREADS be set in platform-indep way in a ctest add_test() command?
